### PR TITLE
enh: openmp to compile matrix multiplication example.

### LINKED ISSUE #4346: compile matrix multiplication example
Towards #3777.

```fortran
program main

!****************************************************************************80
!
!! MAIN is the main program for MXM.
!
!  Licensing:
!
!    This code is distributed under the GNU LGPL license. 
!
!  Modified:
!
!    17 February 2008
!
!  Author:
!
!    John Burkardt
!
  use omp_lib

  implicit none

  integer ( kind = 4 ) id
  integer ( kind = 4 ) l
  integer ( kind = 4 ) m
  integer ( kind = 4 ) n
  integer ( kind = 4 ) thread_num

  print *, ' '
  print *, 'MXM'
  print *, '  FORTRAN90/OpenMP version.'
  print *, ' '
  print *, '  Matrix multiplication tests.'

  print *, ' '
  print *, '  The number of processors available = ', omp_get_num_procs ( )
  print *, '  The number of threads available    = ', omp_get_max_threads ( )

  l = 500
  m = 500
  n = 500

  call r8_mxm ( l, m, n )
!
!  Terminate.
!
  print *, ' '
  print *, 'MXM:'
  print *, '  Normal end of execution.'

  return
end
subroutine r8_mxm ( l, m, n )

!****************************************************************************80
!
!  Purpose:
!
!    R8_MXM carries out a matrix-matrix multiplication in R8 arithmetic.
!
!  Discussion:
!
!    A(LxN) = B(LxM) * C(MxN).
!
!  Licensing:
!
!    This code is distributed under the GNU LGPL license. 
!
!  Modified:
!
!    13 February 2008
!
!  Author:
!
!    John Burkardt
!
!  Parameters:
!
!    Input, integer L, M, N, the dimensions that specify the sizes of the
!    A, B, and C matrices.
!
  use omp_lib

  real    ( kind = 8 ) a(l,n)
  real    ( kind = 8 ) b(l,m)
  real    ( kind = 8 ) c(m,n)
  integer ( kind = 4 ) i
  integer ( kind = 4 ) j
  integer ( kind = 4 ) k
  integer ( kind = 4 ) l
  integer ( kind = 4 ) m
  integer ( kind = 4 ) n
  integer ( kind = 4 ) ops
  real    ( kind = 8 ) r8_uniform_01
  real    ( kind = 8 ) rate
  integer ( kind = 4 ) seed
  real    ( kind = 8 ) time_begin
  real    ( kind = 8 ) time_elapsed
  real    ( kind = 8 ) time_stop
!
!  Assign values to the B and C matrices.
!
  seed = 123456789

  do j = 1, m
    do i = 1, l
      b(i,j) = r8_uniform_01 ( seed )
    end do
  end do

  do j = 1, n
    do i = 1, m
      c(i,j) = r8_uniform_01 ( seed )
    end do
  end do
!
!  Compute A = B * C.
!
  time_begin = omp_get_wtime ( )

!$omp parallel shared(a, b, c, l, m, n) private(i, j, k)

!$omp do
  do j = 1, n
    do i = 1, l
      a(i,j) = 0.0D+00
      do k = 1, m
        a(i,j) = a(i,j) + b(i,k) * c(k,j)
      end do
    end do
  end do
!$omp end do

!$omp end parallel

  time_stop = omp_get_wtime ( )
!
!  Report.
!
  ops = l * n * ( 2 * m )
  time_elapsed = time_stop - time_begin
  rate = dble ( ops ) / time_elapsed / 1000000.0D+00

  print *, ' '
  print *, 'R8_MXM matrix multiplication timing.'
  print *, '  A(LxN) = B(LxM) * C(MxN).'
  print *, '  L = ', l
  print *, '  M = ', m
  print *, '  N = ', n
  print *, '  Floating point OPS roughly ', ops
  print *, '  Elapsed time dT = ', time_elapsed
  print *, '  Rate = MegaOPS/dT = ', rate

  return
end
function r8_uniform_01 ( seed )

!*****************************************************************************80
!
!! R8_UNIFORM_01 returns a unit pseudorandom R8.
!
!  Discussion:
!
!    An R8 is a real ( kind = 8 ) value.
!
!    For now, the input quantity SEED is an integer variable.
!
!    This routine implements the recursion
!
!      seed = 16807 * seed mod ( 2**31 - 1 )
!      r8_uniform_01 = seed / ( 2**31 - 1 )
!
!    The integer arithmetic never requires more than 32 bits,
!    including a sign bit.
!
!    If the initial seed is 12345, then the first three computations are
!
!      Input     Output      R8_UNIFORM_01
!      SEED      SEED
!
!         12345   207482415  0.096616
!     207482415  1790989824  0.833995
!    1790989824  2035175616  0.947702
!
!  Licensing:
!
!    This code is distributed under the GNU LGPL license. 
!
!  Modified:
!
!    05 July 2006
!
!  Author:
!
!    John Burkardt
!
!  Reference:
!
!    Paul Bratley, Bennett Fox, Linus Schrage,
!    A Guide to Simulation,
!    Springer Verlag, pages 201-202, 1983.
!
!    Pierre L'Ecuyer,
!    Random Number Generation,
!    in Handbook of Simulation,
!    edited by Jerry Banks,
!    Wiley Interscience, page 95, 1998.
!
!    Bennett Fox,
!    Algorithm 647:
!    Implementation and Relative Efficiency of Quasirandom
!    Sequence Generators,
!    ACM Transactions on Mathematical Software,
!    Volume 12, Number 4, pages 362-376, 1986.
!
!    Peter Lewis, Allen Goodman, James Miller
!    A Pseudo-Random Number Generator for the System/360,
!    IBM Systems Journal,
!    Volume 8, pages 136-143, 1969.
!
!  Parameters:
!
!    Input/output, integer ( kind = 4 ) SEED, the "seed" value, which should
!    NOT be 0. On output, SEED has been updated.
!
!    Output, real ( kind = 8 ) R8_UNIFORM_01, a new pseudorandom variate,
!    strictly between 0 and 1.
!
  implicit none

  integer ( kind = 4 ), parameter :: i4_huge = 2147483647
  integer ( kind = 4 ) k
  real    ( kind = 8 ) r8_uniform_01
  integer ( kind = 4 ) seed

  if ( seed == 0 ) then
    print *, ' '
    print *, 'R8_UNIFORM_01 - Fatal error!'
    print *, '  Input value of SEED = 0.'
    stop
  end if

  k = seed / 127773

  seed = 16807 * ( seed - k * 127773 ) - k * 2836

  if ( seed < 0 ) then
    seed = seed + i4_huge
  end if
!
!  Although SEED can be represented exactly as a 32 bit integer,
!  it generally cannot be represented exactly as a 32 bit real number!
!
  r8_uniform_01 = real ( seed, kind = 8 ) * 4.656612875D-10

  return
end
```

```console
% lfortran a.f90 
ASR verify pass error: The variable in ArrayItem must be an array, not a scalar
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2478
    LCompilers::print_stack_on_segfault();
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2442
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 912
    result = fe.get_asr2(input, lm, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 256
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 278
    compiler_options.symtab_only, compiler_options);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 107
    if (res.ok) {
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 3811
    b.is_body_visitor = true;
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 144
    visit_ast(*x.m_items[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4878
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4835
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4886
    void visit_program_unit(const program_unit_t &b) { visit_program_unit_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4477
    case program_unitType::Procedure: { v.visit_Procedure((const Procedure_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 2037
    transform_stmts(body, x.n_body, x.m_body);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 115
    this->visit_stmt(*m_body[i]);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4924
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4569
    case stmtType::Assign: { v.visit_Assign((const Assign_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 2352
    ASR::expr_t *target = ASRUtils::EXPR(tmp);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4976
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4626
    case exprType::BoolOp: { v.visit_BoolOp((const BoolOp_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 7389
    this->visit_expr(*x.m_left);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4976
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/ast.h", line 4632
    case exprType::Compare: { v.visit_Compare((const Compare_t &)x); return; }
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 6470
    bool is_function = true;
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 5727
    return resolve_intrinsic_function(x.base.base.loc, var_name);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 6800
    ASR::Module_t *m = ASRUtils::load_module(al, tu_symtab, module_name,
  File "/Users/pranavchiku/repos/lfortran/src/libasr/asr_utils.cpp", line 392
    throw LCompilersException("Verify failed");
LCompilersException: Verify failed
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4346. Towards #3777.

### CODE REVIEW

File: integration_tests/openmp_30.f90 [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+!$omp end parallel do
+end_time = omp_get_wtime()
+
+print *, "Time: ", end_time - start_time
+print *, "sum(a): ", sum(a)
+if (abs(sum(a) - (440952103687207.56D0)) > 1D-12) error stop
+end subroutine
+
+program openmp_30
+call matrix_multiplication(500, 500, 500)
```
Comment (Pranavchiku) [inline integration_tests/openmp_30.f90:32]: We can use this to benchmark our array operations, we are quite close to GFortran even in debug mode. I think we can beat them in release mode.

Comment (Pranavchiku) [reply integration_tests/openmp_30.f90:32]: GFortran ( flags ) - `-O3 -march=native -ffast-math -funroll-loops`
| L=M=N | GFortran | GFortran ( flags ) | LFortran ( --fast ) |
|---|---|---|---|
| 500 | 7.8480999996827450E-002 | 2.1822000002430286E-002 | 2.45089530944824219e-02 |
| 1000 | 0.70620200000121258 | 0.19176799999695504 | 1.88652992248535156e-01 |
| 1500 | 2.4187190000011469 | 0.69105899999703979 | 6.95324182510375977e-01 |
| 2000 | 5.8897070000020904 | 1.8459449999936623 | 1.88580393791198730e+00 |
| 3000 | 24.358389999993960 | 9.1795830000046408 | 8.56550097465515137e+00 |

Comment (certik) [reply integration_tests/openmp_30.f90:32]: Very good. This looks comparable. Now pick the largest L=M=N=4000 and do 1 thread, 2 threads, 4, 8, 16 threads and see. Just to Release mode (`--fast`).

Comment (Pranavchiku) [reply integration_tests/openmp_30.f90:32]: L=M=N=4000
| Num Threads | GFortran ( flags ) | LFortran ( --fast ) |
|---|---|---|
| 1 | 144.33906700000080 | 1.47331972837448120e+02 |
| 2 | 76.052911999999196 | 7.75984909534454346e+01 |
| 4 | 45.320662999998603 | 4.76426410675048828e+01 |
| 8 | 33.889175000000250 | 3.94912590980529785e+01 |
| 16 | 35.262665999998717 | 4.08343570232391357e+01 |
| 32 | 35.644130000000587 | 4.03652579784393311e+01 |

Comment (certik) [reply integration_tests/openmp_30.f90:32]: Very good. We are slightly slower, but competitive. That's great. Thank you!

**Pull request merged**